pkg/vsock/handshake_windows.go

```diff
@@ -48,76 +49,73 @@ func GetVMGUID(seed string, handshakePort uint32, timeout <-chan time.Time) (hvs
 	found := make(chan hvsock.GUID, len(names))
-	done := make(chan bool, len(names))
-	defer close(done)
 
 	for _, name := range names {
 		vmGUID, err := hvsock.GUIDFromString(name)
 		if err != nil {
 			logrus.Errorf("invalid VM name: [%s], err: %v", name, err)
 			continue
 		}
-		go handshake(vmGUID, seed, handshakePort, found, done)
+		go handshake(ctx, vmGUID, handshakePort, seed, found)
 	}
-	return tryFindGUID(found, timeout)
+	return tryFindGUID(cancel, found, timeout)
 }
 
-// handshake attempts to perform a handshake by verifying the seed with a running
-// af_vsock peer in WSL distro, it attempts once per second
-func handshake(vmGUID hvsock.GUID, seedPhrase string, peerHandshakePort uint32, found chan<- hvsock.GUID, done <-chan bool) {
-	svcPort, err := hvsock.GUIDFromString(winio.VsockServiceID(peerHandshakePort).String())
+// GetVsockConnection estabilishes a new AF_VSOCK connection with
+// the provided VM GUID and port, the caller is responsible for closing the connection
+func GetVsockConnection(vmGUID hvsock.GUID, port uint32) (net.Conn, error) {
+	svcPort, err := hvsock.GUIDFromString(winio.VsockServiceID(port).String())
 	if err != nil {
-		logrus.Errorf("hostHandshake parsing svc port: %v", err)
+		return nil, err
 	}
 	addr := hvsock.Addr{
 		VMID:      vmGUID,
 		ServiceID: svcPort,
 	}
 
+	return hvsock.Dial(addr)
+}
+
+// handshake attempts to perform a handshake by verifying the seed with a running
+// af_vsock peer in WSL distro, it attempts once per second
+func handshake(ctx context.Context, vmGUID hvsock.GUID, peerHandshakePort uint32, seedPhrase string, found chan<- hvsock.GUID) {
 	attempInterval := time.NewTicker(time.Second * 1)
-	attempt := 1
+	attempt := 0
 	for {
 		select {
-		case <-done:
+		case <-ctx.Done():
 			logrus.Infof("attempt to handshake with [%s], goroutine is terminated", vmGUID.String())
 			return
 		case <-attempInterval.C:
-			// Spawn a goroutine here to ensure we don't get stuck on a timeout
-			go func() {
-				conn, err := hvsock.Dial(addr)
-				select {
-				case <-done:
-					// If we're already connected, no need to print more things.
-					return
-				default:
-				}
-				if err != nil {
-					attempt++
-					logrus.Debugf("handshake attempt[%v] to dial into VM [%s], looking for vsock-peer", attempt, vmGUID.String())
-					return
-				}
-				seed, err := readSeed(conn, seedPhrase)
-				if err != nil {
-					logrus.Errorf("hosthandshake attempt to read the seed: %v", err)
-				}
-				if err := conn.Close(); err != nil {
-					logrus.Errorf("hosthandshake closing connection: %v", err)
-				}
-				if seed == seedPhrase {
-					logrus.Infof("successfully estabilished a handshake with a peer: %s", vmGUID.String())
-					found <- vmGUID
-					return
-				}
-				logrus.Infof("hosthandshake failed to match the seed phrase with a peer running in: %s", vmGUID.String())
-			}()
+			conn, err := GetVsockConnection(vmGUID, peerHandshakePort)
```

Comment:
`GetVsockConnection` blocks (and can wait for 30 seconds before returning with a time out).  This means that if we're unlucky, we fire a connection request _just_ before the VM is ready, and spend 30 seconds in timeout before the next try — at which point the UI has timed out waiting for the pid file to exist, and tears the whole thing down.  This is why we had that in a goroutine before (so that we can reliably fire off the next attempt one second later, whether or not the previous attempt succeeded).

It turns out I'm pretty good at getting exactly that sort of unlucky, which is why the description is so specific…